sumpy/expansion/__init__.py

```diff
@@ -184,31 +184,68 @@ def copy(self, **kwargs):
 
         return type(self)(**new_kwargs)
 
+    # }}}
+
+    def update_persistent_hash(self, key_hash, key_builder):
+        key_hash.update(type(self).__name__.encode("utf8"))
+        key_builder.rec(key_hash, self.kernel)
+        key_builder.rec(key_hash, self.order)
+        key_builder.rec(key_hash, self.use_rscale)
+
+    def __len__(self):
+        return len(self.get_coefficient_identifiers())
+
+    def __eq__(self, other):
+        return (
+                type(self) == type(other)
+                and self.kernel == other.kernel
+                and self.order == other.order
+                and self.use_rscale == other.use_rscale)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
 
 # }}}
 
 
 # {{{ expansion terms wrangler
 
-class ExpansionTermsWrangler:
+class ExpansionTermsWrangler(ABC):
+    """
+    .. attribute:: order
+    .. attribute:: dim
+    .. attribute:: max_mi
```

Comment:
Type annotations for these?

Reply:
Added some more type annotations to `expansions.__init__` in
https://github.com/inducer/sumpy/compare/e5ed7b5666c6b6b5ab3cfe838ac1f004bec4ef35..a091c5991fc9c2c3e85eff978bba99f700a3d67e